Quadriga/src/main/java/edu/asu/spring/quadriga/web/config/SimpleSignInAdapter.java

```diff
@@ -0,0 +1,122 @@
+import edu.asu.spring.quadriga.domain.IQuadrigaRole;
+import edu.asu.spring.quadriga.domain.IUser;
+import edu.asu.spring.quadriga.exceptions.QuadrigaNotificationException;
+import edu.asu.spring.quadriga.exceptions.QuadrigaStorageException;
+import edu.asu.spring.quadriga.exceptions.UsernameExistsException;
+import edu.asu.spring.quadriga.service.IUserManager;
+import edu.asu.spring.quadriga.service.QuadrigaUserDetails;
+import edu.asu.spring.quadriga.web.config.social.SocialSignInStatus;
+import edu.asu.spring.quadriga.web.login.QuadrigaGrantedAuthority;
+import edu.asu.spring.quadriga.web.manageusers.beans.AccountRequest;
+
+/**
+ * 
+ * This class authorizes (assigns appropriate roles) users authenticated by
+ * Github.
+ * 
+ * @author Chiraag Subramanian
+ *
+ */
+
+public final class SimpleSignInAdapter implements SignInAdapter {
+
+    private final Logger logger = LoggerFactory.getLogger(SignInAdapter.class);
+
+    private IUserManager userManager;
+    private IUserHelper userHelper;
+
+    public SimpleSignInAdapter(IUserManager userManager, IUserHelper userHelper) {
+        this.userManager = userManager;
+        this.userHelper = userHelper;
+    }
+
+    /**
+     * This method checks if Quadriga contains details of the user authenticated
+     * by Github. If the user details are present (which indicates the user is
+     * an authorized user), appropriate roles are assigned, Else if the user
+     * details are not present (which indicates the user is an unregistered /
+     * not approved user, restricted access is assigned)
+     */
+
+    @Override
+    public String signIn(String userId, Connection<?> connection, NativeWebRequest request) {
+        List<GrantedAuthority> authorities = new ArrayList<>();
+        SocialSignInStatus messageType = null;
+        IUser user = (IUser) userManager.findUserByProviderUserId(connection.getKey().getProviderUserId(),
+                connection.getKey().getProviderId());
+        // if user details is not present in the database, create a database
+        // record for the user.
+        if (user == null) {
+            user = userHelper.createUser(connection);
+            AccountRequest accountRequest = new AccountRequest();
+            accountRequest.setUsername(user.getUserName());
+            accountRequest.setName(user.getName());
+            accountRequest.setEmail(user.getEmail());
+            accountRequest.setSocialSignIn(true);
+            accountRequest.setProvider(user.getProvider());
+            accountRequest.setUserIdOfProvider(user.getUserIdOfProvider());
+            messageType = SocialSignInStatus.REGISTRATION_SUCCESS;
+            try {
+                userManager.addNewUser(accountRequest);
+            } catch (QuadrigaStorageException e) {
+                logger.error("Could not add user.", e);
+                messageType = SocialSignInStatus.REGISTRATION_FAILED;
+            } catch (UsernameExistsException e) {
```

Comment:
this is not unexpected and should not be handled through exceptions; check this case first.

Reply:
https://github.com/diging/quadriga/blob/develop/Quadriga/src/main/java/edu/asu/spring/quadriga/web/manageusers/RegisterUserController.java
The usual registration through the form too seems to handle the scenario via UsernameExistsException...

Reply:
that is different. In that case, the method's purpose is to add a new user account, not to sign in. Ideally the validator would check if the user account already exists (it doesn't do that right now, but it should), and the exception would only be used when there are two users trying to use the same username registering at the same time.
Here, this leads to exceptions in the log stating the username is already in use, when the user tries to login with GitHub but the account has not been approved yet, which is very confusing and unnecessary. Would would we try to add another account request when the user tries to login? Especially, since it's just one db call to check if the account request already exists?